src/BuiltInTools/DotNetDeltaApplier/HotReloadAgent.cs

```diff
@@ -45,116 +44,211 @@ private void OnAssemblyLoad(object? _, AssemblyLoadEventArgs eventArgs)
             }
         }
 
-        private sealed class UpdateHandlerActions
+        internal sealed class UpdateHandlerActions
+        {
+            public List<Action<Type[]?>> Before { get; } = new();
+            public List<Action<Type[]?>> After { get; } = new();
+            public List<Action<Type[]?>> ClearCache { get; } = new();
+            public List<Action<Type[]?>> UpdateApplication { get; } = new();
+        }
+
+        private UpdateHandlerActions GetMetadataUpdateHandlerActions()
         {
-            public UpdateHandlerActions(List<Action<Type[]?>> before, List<Action<Type[]?>> after)
+            // In a typical app, a handful of assemblies will have metadata handler attributes. We need to discover these assemblies
+            // and ensure that they are topologically sorted so that handlers in a dependency are executed before the dependent (e.g.
+            // the reflection cache action in System.Private.CoreLib is executed before System.Text.Json clears it's own cache.)
+            // This would ensure that there is a well-defined order and that caches and updates more lower in the application stack
+            // are up to date before ones higher in the stack are recomputed.
+            var assemblies = GetAssembliesWithMetadataUpdateHandlerAttributes();
+            var sortedAssemblies = TopologicalSort(CollectionsMarshal.AsSpan(assemblies));
+            var handlerActions = new UpdateHandlerActions();
+            foreach (var assembly in CollectionsMarshal.AsSpan(sortedAssemblies))
             {
-                Before = before;
-                After = after;
+                foreach (var attribute in assembly.GetCustomAttributes<MetadataUpdateHandlerAttribute>())
```

Comment:
We have said that we are going to match the attribute by name, so that netstandard targeting libraries can define their own copy, without cross-compilation: https://github.com/dotnet/runtime/issues/49361#issuecomment-795992329 . Is it still the plan?

This is matching the attribute by type identity, not by name.

Reply:
Ahh, I might have missed that conversation. I'll switch it back - that's what this code was initially doing before I recently changed it .